### PR TITLE
8332394: Add friendly output when @IR rule missing value

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you help to review this simple patch?
Currently, when a @IR rule like "applyIfPlatform" or "applyIfCPUFeature" miss a value, it will just throw ArrayIndexOutOfBoundsException, with no other information. This is confusing unless you dig into the test frame code.
It's helpful to output more meaningful information.
Thanks

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332394](https://bugs.openjdk.org/browse/JDK-8332394): Add friendly output when @IR rule missing value (**Enhancement** - P4)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [529257fd](https://git.openjdk.org/jdk/pull/19270/files/529257fd2903f9ef886c417848a8a86124045def)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19270/head:pull/19270` \
`$ git checkout pull/19270`

Update a local copy of the PR: \
`$ git checkout pull/19270` \
`$ git pull https://git.openjdk.org/jdk.git pull/19270/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19270`

View PR using the GUI difftool: \
`$ git pr show -t 19270`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19270.diff">https://git.openjdk.org/jdk/pull/19270.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19270#issuecomment-2115342582)

### CODE REVIEW

File: test/hotspot/jtreg/compiler/lib/ir_framework/test/IREncodingPrinter.java [hunk 1]
```diff
@@ -290,6 +290,7 @@ private boolean hasAllRequiredPlatform(String[] andRules) {
         for (int i = 0; i < andRules.length; i++) {
             String platform = andRules[i].trim();
             i++;
+            TestFormat.check(i < andRules.length, "Missing value for platform " + platform + failAt());
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/lib/ir_framework/test/IREncodingPrinter.java:293]: I suggest to also add the `ruleType` as in `hasAllRequiredFlags()`, for example. Then it is even more precise. For even more readability you could add some `""`:
Current:
```
 Missing value for platform xyz in @IR rule 1 at foo()
```
vs.
Improved:
```
 Missing value for platform "xyz" in @IR rule 1 in "applyIfPlatform" at foo()
```

**Review by chhagedorn — APPROVED** [review]
Good catch! Only a small improvement suggestion, otherwise, looks good.

Just noticed that we are actually missing tests that trigger a format violation in  `TestBadFormat` for `applyIfCPUFeature*` and `applyIfPlatform*`. We should probably add some at some point, analogously to the ones already there for `applyIf*` for flags. But that could be done separately.

File: test/hotspot/jtreg/compiler/lib/ir_framework/test/IREncodingPrinter.java (hunk 1 above)
Comment (Hamlin-Li) [reply test/hotspot/jtreg/compiler/lib/ir_framework/test/IREncodingPrinter.java:293]: Thanks, it makes sense.

I also created https://bugs.openjdk.org/browse/JDK-8332402 to track adding tests in `TestBadFormat`

**Comment by Hamlin-Li** [comment]
Thanks @chhagedorn for your reviewing!

/integrate

File: test/hotspot/jtreg/compiler/lib/ir_framework/test/IREncodingPrinter.java (hunk 1 above)
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/lib/ir_framework/test/IREncodingPrinter.java]: Thanks!

Comment (Hamlin-Li) [reply test/hotspot/jtreg/compiler/lib/ir_framework/test/IREncodingPrinter.java]: New pr adding tests is here: https://github.com/openjdk/jdk/pull/19302

**Comment by eme64** [comment]
@Hamlin-Li you only got 1 review. Per the rules, you generally need 2:
https://openjdk.org/guide/#final-check-before-creating-the-pr

That is unless you say that the change is **trivial**, and the reviewer also confirms that it is **trivial**. I don't see that here.

Our rule is that you need 2 reviewers: at least one reviewer, the second one can be a committer.

**Pull request closed without merging**